### PR TITLE
Make the node label for gluster pods configurable

### PR DESCRIPTION
**Pull request by MohamedAshiqrh**
Signed-off-by: Mohamed Ashiq Liyazudeen <mliyazud@redhat.com>

---
This change is [Reviewable](https://reviewable.io/reviews/gluster/gluster-kubernetes/255)

### CODE REVIEW

File: deploy/gk-deploy [hunk 1]
```diff
@@ -31,11 +31,12 @@ SKIP_PREREQ=0
 LOAD=0
 ADMIN_KEY=""
 USER_KEY=""
+GLUSTER_LABEL=""
 
 usage() {
   echo -e "USAGE: ${PROG} [-ghvy] [-c CLI] [-t <TEMPLATES>] [-n NAMESPACE] [-w <SECONDS>]
        [--load] [--admin-key <ADMIN_KEY>] [--user-key <USER_KEY>] [-l <LOG_FILE>]
-       [<TOPOLOGY>]\n"
+       [--gluster-label <GLUSTER_LABEL> ]  [<TOPOLOGY>]\n"
```
Comment (humblec) [inline deploy/gk-deploy]: May be 'node-selector' or similar would be a good name for this option. Gluster label is bit confusing as per me.

**Comment by humblec** [comment]
@MohamedAshiqrh Thanks! Overall patch looks good to me.  Minor comment from me though. Also did you get a chance to test this in multiple namespace and also with different cns-deploy iterations to make sure all scenarios works perfectly?

File: deploy/gk-deploy (hunk 1 above)
Comment (jarrpa) [inline deploy/gk-deploy]: I agree, `--daemonset-label` would be better.

File: deploy/gk-deploy [hunk 2]
```diff
@@ -88,6 +89,11 @@ Options:
               APIs. user key: string, Shared secret.
               Eg: MyVolumeSecret
 
+  --gluster-label GLUSTER_LABEL
+              The storage-node label set on openshift nodes are used to select the
+              place where gluster pods should be started. To have multiple cluster
+              we need to use different name. This flag will set the user defined label.
+
```
Comment (jarrpa) [inline deploy/gk-deploy]: The above should be changed to:
```
            Controls the value of the label set on nodes which will host pods
            from the GlusterFS daemonset. This allows for multiple GlusterFS
            daemonsets to run in the same cluster. Default is 'glusterfs'.
```

File: deploy/ocp-templates/glusterfs-template.yaml [hunk 3]
```diff
@@ -119,3 +119,8 @@ objects:
         terminationGracePeriodSeconds: 30
         dnsPolicy: ClusterFirst
         securityContext: {}
+parameters:
+- name: NODE_LABEL
+  displayName: Gluster Node Label
+  description: Allows user to modify the label to be used to select node.
```
Comment (jarrpa) [inline deploy/ocp-templates/glusterfs-template.yaml]: The above should be changed to:
```
displayName: Daemonset Storage Node Label
description: This is the value that will be used for the storagenode label in the daemonset node selector.
```

File: deploy/gk-deploy [hunk 4]
```diff
@@ -468,20 +479,22 @@ if [[ ${LOAD} -eq 0 ]]; then
...
       abort
     fi
   done <<< "$(echo -e "${NODES}")"
   debug "Deploying GlusterFS pods."
   if [[ "${CLI}" == *oc\ * ]]; then
-    eval_output "${CLI} process glusterfs | ${CLI} create -f - 2>&1"
+    eval_output "${CLI} process ${OC_PROCESS_VAL_SWITCH} NODE_LABEL=${GLUSTER_LABEL} glusterfs | ${CLI} create -f - 2>&1"
   else
-    eval_output "${CLI} create -f ${TEMPLATES}/glusterfs-daemonset.yaml 2>&1"
+    eval_output "sed -e 's/storagenode\: glusterfs/storagenode\: '"${GLUSTER_LABEL}"'/g' ${TEMPLATES}/glusterfs-daemonset.yaml | ${CLI} create -f - 2>&1"
```
Comment (obnoxxx) [inline deploy/gk-deploy]: There is a quoting error here, found by make test...

Comment (MohamedAshiqrh) [inline deploy/gk-deploy]: taken care of. @obnoxxx ++ for the tests. Feels more confident about the change.

File: deploy/ocp-templates/glusterfs-template.yaml (hunk 3 above)
Comment (MohamedAshiqrh) [inline deploy/ocp-templates/glusterfs-template.yaml]: Done.

File: deploy/gk-deploy (hunk 2 above)
Comment (MohamedAshiqrh) [inline deploy/gk-deploy]: Done. Thanks this is more clearer.

File: deploy/gk-deploy (hunk 1 above)
Comment (MohamedAshiqrh) [inline deploy/gk-deploy]: Thanks done.

**Comment by MohamedAshiqrh** [comment]
```
# gk-deploy --daemonset-label glusterfs2 -g topology1.json 
Welcome to the deployment tool for GlusterFS on Kubernetes and OpenShift.

Before getting started, this script has some requirements of the execution
environment and of the container platform that you should verify.

The client machine that will run this script must have:
 * Administrative access to an existing Kubernetes or OpenShift cluster
 * Access to a python interpreter 'python'
 * Access to the heketi client 'heketi-cli'

Each of the nodes that will host GlusterFS must also have appropriate firewall
rules for the required GlusterFS ports:
 * 2222  - sshd (if running GlusterFS in a pod)
 * 24007 - GlusterFS Daemon
 * 24008 - GlusterFS Management
 * 49152 to 49251 - Each brick for every volume on the host requires its own
   port. For every new brick, one new port will be used starting at 49152. We
   recommend a default range of 49152-49251 on each host, though you can adjust
   this to fit your needs.

In addition, for an OpenShift deployment you must:
 * Have 'cluster_admin' role on the administrative account doing the deployment
 * Add the 'default' and 'router' Service Accounts to the 'privileged' SCC
 * Have a router deployed that is configured to allow apps to access services
   running in the cluster

Do you wish to proceed with deployment?

[Y]es, [N]o? [Default: Y]: Y
Using OpenShift CLI.
NAME                STATUS    AGE
storage-project-2   Active    2m
Using namespace "storage-project-2".
Checking that heketi pod is not running ... OK
template "deploy-heketi" created
serviceaccount "heketi-service-account" created
template "heketi" created
template "glusterfs" created
role "edit" added: "system:serviceaccount:storage-project-2:heketi-service-account"
node "192.168.35.5" labeled
node "192.168.35.6" labeled
node "192.168.35.7" labeled
daemonset "glusterfs" created
Waiting for GlusterFS pods to start ... OK
service "deploy-heketi" created
route "deploy-heketi" created
deploymentconfig "deploy-heketi" created
Waiting for deploy-heketi pod to start ... OK
Creating cluster ... ID: fde139c21b0afcb6206bf272e0df1590
Creating node 192.168.35.5 ... ID: 0768a1ee35dce4cf707c7a1e9caa3d2a
Adding device /dev/vdc ... OK
Adding device /dev/vdd ... OK
Adding device /dev/vde ... OK
Adding device /dev/vdf ... OK
Creating node 192.168.35.6 ... ID: 63966f6ffd48c1980c4a2d03abeedd04
Adding device /dev/vdc ... OK
Adding device /dev/vdd ... OK
Adding device /dev/vde ... OK
Adding device /dev/vdf ... OK
Creating node 192.168.35.7 ... ID: de129c099193aaff2c64dca825f33558
Adding device /dev/vdc ... OK
Adding device /dev/vdd ... OK
Adding device /dev/vde ... OK
Adding device /dev/vdf ... OK
heketi topology loaded.
Saving heketi-storage.json
secret "heketi-storage-secret" created
endpoints "heketi-storage-endpoints" created
service "heketi-storage-endpoints" created
job "heketi-storage-copy-job" created
deploymentconfig "deploy-heketi" deleted
route "deploy-heketi" deleted
service "deploy-heketi" deleted
job "heketi-storage-copy-job" deleted
pod "deploy-heketi-1-d0qrs" deleted
secret "heketi-storage-secret" deleted
service "heketi" created
route "heketi" created
deploymentconfig "heketi" created
Waiting for heketi pod to start ... OK
heketi is now running.
Ready to create and provide GlusterFS volumes.
```

> # oc get ds
> NAME        DESIRED   CURRENT   READY     NODE-SELECTOR            AGE
> glusterfs   3         3         3         storagenode=glusterfs2   8m
> 


I had gk-deploy on storage-project and this was on other namespace which I choose a new label for daemonset.

**Comment by MohamedAshiqrh** [comment]
@humblec @obnoxxx @jarrpa PTAL. Tested it on my environment .

**Comment by jarrpa** [comment]
LGTM! If we missed anything over three reviewers, we can probably live with fixing it in another PR. ;) Merging.

**Pull request merged**